File: src/main/java/net/atos/zac/policy/PolicyService.java
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2024 Lifely
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.zac.policy;

import static net.atos.client.zgw.drc.model.generated.StatusEnum.DEFINITIEF;
import static net.atos.client.zgw.util.UriUtilsKt.extractUuid;
import static net.atos.client.zgw.zrc.util.StatusTypeUtil.isHeropend;
import static net.atos.client.zgw.zrc.util.StatusTypeUtil.isIntake;
import static net.atos.zac.enkelvoudiginformatieobject.util.EnkelvoudigInformatieObjectCheckersKt.isSigned;
import static net.atos.zac.flowable.task.TaakVariabelenService.readZaaktypeOmschrijving;
import static net.atos.zac.flowable.util.TaskUtil.isOpen;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.enterprise.inject.Instance;
import jakarta.inject.Inject;

import org.apache.commons.collections.CollectionUtils;
import org.eclipse.microprofile.rest.client.inject.RestClient;
import org.flowable.task.api.TaskInfo;

import net.atos.client.opa.model.RuleQuery;
import net.atos.client.zgw.drc.model.generated.EnkelvoudigInformatieObject;
import net.atos.client.zgw.zrc.ZrcClientService;
import net.atos.client.zgw.zrc.model.Zaak;
import net.atos.client.zgw.ztc.ZtcClientService;
import net.atos.client.zgw.ztc.model.generated.StatusType;
import net.atos.client.zgw.ztc.model.generated.ZaakType;
import net.atos.zac.authentication.LoggedInUser;
import net.atos.zac.enkelvoudiginformatieobject.EnkelvoudigInformatieObjectLockService;
import net.atos.zac.enkelvoudiginformatieobject.model.EnkelvoudigInformatieObjectLock;
import net.atos.zac.policy.exception.PolicyException;
import net.atos.zac.policy.input.DocumentData;
import net.atos.zac.policy.input.DocumentInput;
import net.atos.zac.policy.input.TaakData;
import net.atos.zac.policy.input.TaakInput;
import net.atos.zac.policy.input.UserInput;
import net.atos.zac.policy.input.ZaakData;
import net.atos.zac.policy.input.ZaakInput;
import net.atos.zac.policy.output.DocumentRechten;
import net.atos.zac.policy.output.OverigeRechten;
import net.atos.zac.policy.output.TaakRechten;
import net.atos.zac.policy.output.WerklijstRechten;
import net.atos.zac.policy.output.ZaakRechten;
import net.atos.zac.shared.exception.FoutmeldingException;
import net.atos.zac.zoeken.model.DocumentIndicatie;
import net.atos.zac.zoeken.model.ZaakIndicatie;
import net.atos.zac.zoeken.model.zoekobject.DocumentZoekObject;
import net.atos.zac.zoeken.model.zoekobject.TaakZoekObject;
import net.atos.zac.zoeken.model.zoekobject.ZaakZoekObject;

@ApplicationScoped
public class PolicyService {
    private Instance<LoggedInUser> loggedInUserInstance;
    private OPAEvaluationClient evaluationClient;
    private ZtcClientService ztcClientService;
    private EnkelvoudigInformatieObjectLockService lockService;
    private ZrcClientService zrcClientService;

    /**
     * Default no-arg constructor, required by Weld.
     */
    public PolicyService() {
    }

    @Inject
    public PolicyService(
            final Instance<LoggedInUser> loggedInUserInstance,
            final @RestClient OPAEvaluationClient evaluationClient,
            final ZtcClientService ztcClientService,
            final EnkelvoudigInformatieObjectLockService lockService,
            final ZrcClientService zrcClientService
    ) {
        this.loggedInUserInstance = loggedInUserInstance;
        this.evaluationClient = evaluationClient;
        this.ztcClientService = ztcClientService;
        this.lockService = lockService;
        this.zrcClientService = zrcClientService;
    }

    public OverigeRechten readOverigeRechten() {
        return evaluationClient.readOverigeRechten(
                new RuleQuery<>(new UserInput(loggedInUserInstance.get()))
        ).getResult();
    }

    public ZaakRechten readZaakRechten(final Zaak zaak) {
        return readZaakRechten(zaak, ztcClientService.readZaaktype(zaak.getZaaktype()));
    }

    public ZaakRechten readZaakRechten(final Zaak zaak, final ZaakType zaaktype) {
        final ZaakData zaakData = new ZaakData();
        zaakData.open = zaak.isOpen();
        zaakData.zaaktype = zaaktype.getOmschrijving();
        zaakData.opgeschort = zaak.isOpgeschort();
        zaakData.verlengd = zaak.isVerlengd();
        zaakData.besloten = CollectionUtils.isNotEmpty(zaaktype.getBesluittypen());
        StatusType statusType = null;
        if (zaak.getStatus() != null) {
            var status = zrcClientService.readStatus(zaak.getStatus());
            statusType = ztcClientService.readStatustype(status.getStatustype());
        }
        zaakData.intake = isIntake(statusType);
        zaakData.heropend = isHeropend(statusType);
        return evaluationClient.readZaakRechten(new RuleQuery<>(
                new ZaakInput(loggedInUserInstance.get(), zaakData))
        ).getResult();
    }

    public ZaakRechten readZaakRechten(final ZaakZoekObject zaakZoekObject) {
        final ZaakData zaakData = new ZaakData();
        zaakData.open = !zaakZoekObject.isAfgehandeld();
        zaakData.zaaktype = zaakZoekObject.getZaaktypeOmschrijving();
        zaakData.opgeschort = zaakZoekObject.getZaakIndicaties().contains(ZaakIndicatie.OPSCHORTING);
        zaakData.verlengd = zaakZoekObject.getZaakIndicaties().contains(ZaakIndicatie.VERLENGD);
        zaakData.heropend = zaakZoekObject.getZaakIndicaties().contains(ZaakIndicatie.HEROPEND);

        return evaluationClient.readZaakRechten(new RuleQuery<>(new ZaakInput(loggedInUserInstance.get(), zaakData)))
                .getResult();
    }

    public DocumentRechten readDocumentRechten(final EnkelvoudigInformatieObject enkelvoudigInformatieobject) {
        return readDocumentRechten(enkelvoudigInformatieobject, null);
    }

    public DocumentRechten readDocumentRechten(final EnkelvoudigInformatieObject enkelvoudigInformatieobject, final Zaak zaak) {
        return readDocumentRechten(
                enkelvoudigInformatieobject,
                lockService.findLock(extractUuid(enkelvoudigInformatieobject.getUrl())),
                zaak
        );
    }

    public DocumentRechten readDocumentRechten(
            final EnkelvoudigInformatieObject enkelvoudigInformatieobject,
            final EnkelvoudigInformatieObjectLock lock,
            final Zaak zaak
    ) {
        final DocumentData documentData = new DocumentData();
        documentData.definitief = enkelvoudigInformatieobject.getStatus() == DEFINITIEF;
        documentData.vergrendeld = enkelvoudigInformatieobject.getLocked();
        documentData.vergrendeldDoor = lock != null ? lock.getUserId() : null;
        documentData.ondertekend = isSigned(enkelvoudigInformatieobject);
        if (zaak != null) {
            documentData.zaakOpen = zaak.isOpen();
            documentData.zaaktype = ztcClientService.readZaaktype(zaak.getZaaktype()).getOmschrijving();
        }
        return evaluationClient.readDocumentRechten(
                new RuleQuery<>(new DocumentInput(loggedInUserInstance.get(), documentData))).getResult();
    }

    public DocumentRechten readDocumentRechten(final DocumentZoekObject enkelvoudigInformatieobject) {
        final DocumentData documentData = new DocumentData();
        documentData.definitief = DEFINITIEF.equals(enkelvoudigInformatieobject.getStatus());
        documentData.vergrendeld = enkelvoudigInformatieobject.isIndicatie(DocumentIndicatie.VERGRENDELD);
        documentData.vergrendeldDoor = enkelvoudigInformatieobject.getVergrendeldDoorGebruikersnaam();
        documentData.zaakOpen = !enkelvoudigInformatieobject.isZaakAfgehandeld();
        documentData.zaaktype = enkelvoudigInformatieobject.getZaaktypeOmschrijving();
        documentData.ondertekend = enkelvoudigInformatieobject.getOndertekeningDatum() != null;
        return evaluationClient.readDocumentRechten(
                new RuleQuery<>(new DocumentInput(loggedInUserInstance.get(), documentData))).getResult();
    }

    public TaakRechten readTaakRechten(final TaskInfo taskInfo) {
        return readTaakRechten(taskInfo, readZaaktypeOmschrijving(taskInfo));
    }

    public TaakRechten readTaakRechten(final TaskInfo taskInfo, final String zaaktypeOmschrijving) {
        final TaakData taakData = new TaakData();
        taakData.open = isOpen(taskInfo);
        taakData.zaaktype = zaaktypeOmschrijving;
        return evaluationClient.readTaakRechten(new RuleQuery<>(new TaakInput(loggedInUserInstance.get(), taakData)))
                .getResult();
    }

    public TaakRechten readTaakRechten(final TaakZoekObject taakZoekObject) {
        final TaakData taakData = new TaakData();
        taakData.zaaktype = taakZoekObject.getZaaktypeOmschrijving();
        return evaluationClient.readTaakRechten(
                new RuleQuery<>(new TaakInput(loggedInUserInstance.get(), taakData))
        ).getResult();
    }

    public WerklijstRechten readWerklijstRechten() {
        return evaluationClient.readWerklijstRechten(
                new RuleQuery<>(new UserInput(loggedInUserInstance.get()))
        ).getResult();
    }

    public void checkZaakAfsluitbaar(final Zaak zaak) {
        if (zrcClientService.heeftOpenDeelzaken(zaak)) {
            throw new FoutmeldingException("Deze hoofdzaak heeft open deelzaken en kan niet afgesloten worden.");
        }
        if (lockService.hasLockedInformatieobjecten(zaak)) {
            throw new FoutmeldingException("Deze zaak heeft vergrendelde documenten en kan niet afgesloten worden.");
        }
    }

    public static void assertPolicy(final boolean policy) {
        if (!policy) {
            throw new PolicyException();
        }
    }
}

File: src/main/kotlin/nl/info/zac/policy/PolicyService.kt
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2024 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.zac.policy

import jakarta.enterprise.context.ApplicationScoped
import jakarta.enterprise.inject.Instance
import jakarta.inject.Inject
import net.atos.zac.flowable.task.TaakVariabelenService
import net.atos.zac.flowable.util.TaskUtil
import nl.info.client.opa.model.RuleQuery
import nl.info.client.zgw.drc.model.generated.EnkelvoudigInformatieObject
import nl.info.client.zgw.drc.model.generated.StatusEnum
import nl.info.client.zgw.util.extractUuid
import nl.info.client.zgw.zrc.ZrcClientService
import nl.info.client.zgw.zrc.model.generated.Zaak
import nl.info.client.zgw.zrc.util.isHeropend
import nl.info.client.zgw.zrc.util.isIntake
import nl.info.client.zgw.zrc.util.isOpen
import nl.info.client.zgw.zrc.util.isOpgeschort
import nl.info.client.zgw.zrc.util.isVerlengd
import nl.info.client.zgw.ztc.ZtcClientService
import nl.info.client.zgw.ztc.model.generated.ZaakType
import nl.info.zac.authentication.LoggedInUser
import nl.info.zac.enkelvoudiginformatieobject.EnkelvoudigInformatieObjectLockService
import nl.info.zac.enkelvoudiginformatieobject.model.EnkelvoudigInformatieObjectLock
import nl.info.zac.enkelvoudiginformatieobject.util.isSigned
import nl.info.zac.policy.exception.PolicyException
import nl.info.zac.policy.input.DocumentData
import nl.info.zac.policy.input.DocumentInput
import nl.info.zac.policy.input.TaakData
import nl.info.zac.policy.input.TaakInput
import nl.info.zac.policy.input.UserInput
import nl.info.zac.policy.input.ZaakData
import nl.info.zac.policy.input.ZaakInput
import nl.info.zac.policy.output.DocumentRechten
import nl.info.zac.policy.output.NotitieRechten
import nl.info.zac.policy.output.OverigeRechten
import nl.info.zac.policy.output.TaakRechten
import nl.info.zac.policy.output.WerklijstRechten
import nl.info.zac.policy.output.ZaakRechten
import nl.info.zac.search.model.DocumentIndicatie
import nl.info.zac.search.model.ZaakIndicatie
import nl.info.zac.search.model.zoekobject.DocumentZoekObject
import nl.info.zac.search.model.zoekobject.TaakZoekObject
import nl.info.zac.search.model.zoekobject.ZaakZoekObject
import nl.info.zac.util.AllOpen
import nl.info.zac.util.NoArgConstructor
import org.eclipse.microprofile.rest.client.inject.RestClient
import org.flowable.task.api.TaskInfo

@ApplicationScoped
@NoArgConstructor
@AllOpen
@Suppress("TooManyFunctions")
class PolicyService @Inject constructor(
    private val loggedInUserInstance: Instance<LoggedInUser>,
    @RestClient private val evaluationClient: OpaEvaluationClient,
    private val ztcClientService: ZtcClientService,
    private val lockService: EnkelvoudigInformatieObjectLockService,
    private val zrcClientService: ZrcClientService
) {
    fun readOverigeRechten(): OverigeRechten =
        readOverigeRechten(zaaktype = null)

    // for PABC based IAM integration
    fun readOverigeRechten(zaaktype: String?): OverigeRechten =
        evaluationClient.readOverigeRechten(
            RuleQuery(UserInput(loggedInUserInstance.get(), zaaktype))
        ).result

    fun readZaakRechten(zaak: Zaak): ZaakRechten {
        val zaakType = ztcClientService.readZaaktype(zaak.zaaktype)
        return readZaakRechten(zaak, zaakType)
    }

    fun readZaakRechten(zaak: Zaak, zaaktype: ZaakType): ZaakRechten {
        val statusType = zaak.status?.let {
            ztcClientService.readStatustype(zrcClientService.readStatus(it).statustype)
        }
        val zaakData = ZaakData(
            open = zaak.isOpen(),
            zaaktype = zaaktype.getOmschrijving(),
            opgeschort = zaak.isOpgeschort(),
            verlengd = zaak.isVerlengd(),
            besloten = zaaktype.getBesluittypen()?.isNotEmpty() == true,
            intake = statusType?.isIntake(),
            heropend = statusType?.isHeropend()
        )
        return evaluationClient.readZaakRechten(
            RuleQuery(
                ZaakInput(loggedInUserInstance.get(), zaakData)
            )
        ).result
    }

    fun readZaakRechtenForZaakZoekObject(zaakZoekObject: ZaakZoekObject): ZaakRechten {
        val zaakData = ZaakData(
            open = !zaakZoekObject.isAfgehandeld,
            zaaktype = zaakZoekObject.zaaktypeOmschrijving,
            opgeschort = zaakZoekObject.getZaakIndicaties().contains(ZaakIndicatie.OPSCHORTING),
            verlengd = zaakZoekObject.getZaakIndicaties().contains(ZaakIndicatie.VERLENGD),
            heropend = zaakZoekObject.getZaakIndicaties().contains(ZaakIndicatie.HEROPEND),
            // not taken into account when searching for a zaak
            intake = null,
            // not taken into account when searching for a zaak
            besloten = null
        )
        return evaluationClient.readZaakRechten(
            RuleQuery(ZaakInput(loggedInUserInstance.get(), zaakData))
        ).result
    }

    fun readDocumentRechten(enkelvoudigInformatieobject: EnkelvoudigInformatieObject, zaak: Zaak? = null) =
        readDocumentRechten(
            enkelvoudigInformatieobject = enkelvoudigInformatieobject,
            lock = lockService.findLock(enkelvoudigInformatieobject.getUrl().extractUuid()),
            zaak = zaak
        )

    fun readDocumentRechten(
        enkelvoudigInformatieobject: EnkelvoudigInformatieObject,
        lock: EnkelvoudigInformatieObjectLock?,
        zaak: Zaak?
    ): DocumentRechten {
        val documentData = DocumentData(
            definitief = enkelvoudigInformatieobject.getStatus() == StatusEnum.DEFINITIEF,
            vergrendeld = enkelvoudigInformatieobject.getLocked(),
            vergrendeldDoor = lock?.userId,
            ondertekend = enkelvoudigInformatieobject.isSigned(),
            zaakOpen = zaak?.isOpen() ?: false,
            zaaktype = zaak?.let { ztcClientService.readZaaktype(it.getZaaktype()).getOmschrijving() }
        )
        return evaluationClient.readDocumentRechten(
            RuleQuery(DocumentInput(loggedInUserInstance.get(), documentData))
        ).result
    }

    fun readDocumentRechten(enkelvoudigInformatieobject: DocumentZoekObject): DocumentRechten {
        val documentData = DocumentData(
            definitief = StatusEnum.DEFINITIEF == enkelvoudigInformatieobject.getStatus(),
            vergrendeld = enkelvoudigInformatieobject.isIndicatie(DocumentIndicatie.VERGRENDELD),
            vergrendeldDoor = enkelvoudigInformatieobject.vergrendeldDoorGebruikersnaam,
            zaakOpen = !enkelvoudigInformatieobject.isZaakAfgehandeld,
            zaaktype = enkelvoudigInformatieobject.zaaktypeOmschrijving,
            ondertekend = enkelvoudigInformatieobject.ondertekeningDatum != null
        )
        return evaluationClient.readDocumentRechten(
            RuleQuery(DocumentInput(loggedInUserInstance.get(), documentData))
        ).result
    }

    fun readTaakRechten(taskInfo: TaskInfo): TaakRechten {
        val zaaktypeOmschrijving = TaakVariabelenService.readZaaktypeOmschrijving(taskInfo)
        return readTaakRechten(taskInfo, zaaktypeOmschrijving)
    }

    fun readTaakRechten(
        taskInfo: TaskInfo,
        zaaktypeOmschrijving: String?
    ): TaakRechten {
        val taakData = TaakData(
            open = TaskUtil.isOpen(taskInfo),
            zaaktype = zaaktypeOmschrijving
        )
        return evaluationClient.readTaakRechten(
            RuleQuery(TaakInput(loggedInUserInstance.get(), taakData))
        ).result
    }

    fun readTaakRechten(taakZoekObject: TaakZoekObject): TaakRechten {
        val taakData = TaakData(
            zaaktype = taakZoekObject.zaaktypeOmschrijving
        )
        return evaluationClient.readTaakRechten(
            RuleQuery(TaakInput(loggedInUserInstance.get(), taakData))
        ).result
    }

    fun readNotitieRechten(): NotitieRechten =
        evaluationClient.readNotitieRechten(
            RuleQuery(UserInput(loggedInUserInstance.get()))
        ).result

    fun readWerklijstRechten(): WerklijstRechten =
        evaluationClient.readWerklijstRechten(
            RuleQuery(UserInput(loggedInUserInstance.get()))
        ).result
}

/**
 * Assert that the given policy is true.
 * If it is not, throw a [PolicyException].
 */
fun assertPolicy(policy: Boolean) {
    if (!policy) {
        throw PolicyException()
    }
}
